In [1]:
import json 
import pandas as pd
import numpy as np
from pandas.io.json import json_normalize #package for flattening json in pandas df
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.decomposition import LatentDirichletAllocation

In [2]:
df = pd.read_json(r'/opt/data/articles_list_Oct12_2020.json', orient='columns')

#### Importing Source-Lean-Flags.csv

In [11]:
df_src_flags=pd.read_csv('source-lean-flags.csv')

In [12]:
df_src_flags.describe()

bias_score  factually_questionable_flag  propaganda_flag  \
count  2723.000000                  3050.000000      3050.000000   
mean      0.371282                     0.273443         0.028852   
std       1.762916                     0.445799         0.167419   
min      -3.000000                     0.000000         0.000000   
25%      -1.000000                     0.000000         0.000000   
50%       0.000000                     0.000000         0.000000   
75%       2.000000                     1.000000         0.000000   
max       3.000000                     1.000000         1.000000   

       conspiracy_flag  hate_group_flag  satire_flag  
count      3050.000000      3050.000000  3050.000000  
mean          0.096066         0.022623     0.038033  
std           0.294729         0.148723     0.191307  
min           0.000000         0.000000     0.000000  
25%           0.000000         0.000000     0.000000  
50%           0.000000         0.000000     0.000000  
75%           0.000000         0.000000     0.000000  
max           1.000000         1.000000     1.000000

In [13]:
df_src_flags.count()

domain                         3050
bias_score                     2723
factually_questionable_flag    3050
propaganda_flag                3050
conspiracy_flag                3050
hate_group_flag                3050
satire_flag                    3050
dtype: int64

In [15]:
df_src_flags.shape

(3050, 7)

## Basic EDA

In [3]:
df.describe()

_score
count  60374.0
mean       1.0
std        0.0
min        1.0
25%        1.0
50%        1.0
75%        1.0
max        1.0

In [4]:
df.count()

_index     60374
_type      60374
_id        60374
_score     60374
_source    60374
dtype: int64

In [16]:
df.shape

(60374, 5)

In [6]:
#print 1st column 1st row
df.iloc[0] # this is same as df[[0]].values[0][0]

_index                                              articles
_type                                                   _doc
_id                                     jzHWHnUBlmHqFksVvSCK
_score                                                     1
_source    {'url': 'http://actionnewsjax.com/news/trendin...
Name: 0, dtype: object

In [7]:
df_clean= df.drop(['_id','_score', '_type'], axis=1)

In [30]:
df_clean.shape

(60374, 2)

In [18]:
df_clean.head(5)

_index                                            _source
0  articles  {'url': 'http://actionnewsjax.com/news/trendin...
1  articles  {'url': 'https://www.13newsnow.com/article/new...
2  articles  {'url': 'http://digg.com/digg-picks/link/kindl...
3  articles  {'url': 'http://sports.mynorthwest.com/1199798...
4  articles  {'url': 'https://www.aier.org/article/lockdown...

In [17]:
df_clean.iloc[0]

_index                                              articles
_source    {'url': 'http://actionnewsjax.com/news/trendin...
Name: 0, dtype: object

In [19]:
df.iloc[0]

_index                                              articles
_type                                                   _doc
_id                                     jzHWHnUBlmHqFksVvSCK
_score                                                     1
_source    {'url': 'http://actionnewsjax.com/news/trendin...
Name: 0, dtype: object

In [20]:
df.describe()

_score
count  60374.0
mean       1.0
std        0.0
min        1.0
25%        1.0
50%        1.0
75%        1.0
max        1.0

In [21]:
df.dtypes

_index     object
_type      object
_id        object
_score      int64
_source    object
dtype: object

In [22]:
#Get value within a Series
df.loc[0].at['_type']

'_doc'

In [44]:
#Get value within a Series
df.loc[0].at['_source']

{'url': 'http://actionnewsjax.com/news/trending/amy-coney-barrett-hearing-day-2-questioning-begins-livestream-live-updates/MSRUGTDZVRFO3M7PXIQC6UY45U/',
 'extracted': {'title': 'Amy Coney Barrett hearing Day 2: Questioning begins; livestream, live updates',
  'date': None,
  'authors': ['Debbie Lord',
   'Cox Media Group National Content Desk',
   'Updated',
   'October',
   '- Pm',
   'Kelli Dugan',
   'Crystal Bonvillian',
   'Brianna Chambers',
   'Theresa Seiger'],
  'text': '“His judicial philosophy was straightforward: A judge must apply the law as written, not as the judge wishes it were. Sometimes that approach meant reaching results that he did not like. But as he put it in one of his best-known opinions, that is what it means to say we have a government of laws, not of men.”'},
 'metadata': {'keywords': ['amy coney barrett',
   'amy coney barrett hearing',
   'what time doe the amy coney barrett hearing start',
   'what channel is the amy coney barrett hearing on',
   'livest

In [114]:
df_src=df.drop(['_index','_id','_score', '_type'], axis=1)

In [115]:
df_src.dtypes

_source    object
dtype: object

In [116]:
type(df_src)

pandas.core.frame.DataFrame

In [117]:
%%time
df_src_norm = json_normalize(df_src['_source'])

CPU times: user 51.5 s, sys: 661 ms, total: 52.2 s
Wall time: 52.1 s


In [118]:
df_src_norm.shape

(60374, 2481)

In [113]:
# List all column names
#list(df_src_norm.columns) 
#print(df_src_norm.columns.tolist())

In [45]:
df_src_norm.loc[0]

url                                                                   http://actionnewsjax.com/news/trending/amy-con...
extracted.title                                                       Amy Coney Barrett hearing Day 2: Questioning b...
extracted.date                                                                                                     None
extracted.authors                                                     [Debbie Lord, Cox Media Group National Content...
extracted.text                                                        “His judicial philosophy was straightforward: ...
                                                                                            ...                        
metadata.dump.webit_cover_date                                                                                      NaN
metadata.dump.burntime                                                                                              NaN
metadata.dump.Everything we know about A

In [102]:
#domain_cols = [col for col in df_src_norm.columns if 'domain' in col]
#print(list(df.columns))
#print(domain_cols)
#type(domain_cols)
#df_domain = df_src_norm[domain_cols]

['metadata.dump.twitter.domain', 'metadata.dump.facebook-domain-verification', 'metadata.dump.p.domain_verify', 'metadata.dump.dailymotion-domain-verification', 'metadata.dump.og.domain', 'metadata.dump.com.silverpop.brandeddomains', 'metadata.dump.update-cart-domain', 'metadata.dump.preview-new-proration-domain', 'metadata.dump.preview-change-proration-domain', 'metadata.dump.sso-api-domain', 'metadata.dump.revv-api-domain', 'metadata.dump.analytics.site_domain', 'metadata.dump.globalsign-domain-verification', 'metadata.dump.zoom-domain-verification', 'metadata.dump.cna.subdomain_id', 'metadata.dump.rand-domains', 'metadata.dump.domain', 'metadata.dump.“p.domain_verify”', 'metadata.dump.dfp.domain']


In [103]:
#df_domain.dropna(how='all')

metadata.dump.twitter.domain metadata.dump.facebook-domain-verification  \
1                www.13newsnow.com                                        NaN   
4                     www.aier.org                                        NaN   
5                           rte.ie                                        NaN   
13                             NaN             it3qhk73o3gnhiw5gnatg2k4vs32ks   
16                             NaN                                        NaN   
...                            ...                                        ...   
60358                          NaN                                        NaN   
60360                          NaN                                        NaN   
60367                          NaN                                        NaN   
60371                          NaN                                        NaN   
60372                  twitchy.com                                        NaN   

          metadata.dump.p.domain_verify  \
1                                   NaN   
4                                   NaN   
5                                   NaN   
13                                  NaN   
16                    69735562919053270   
...                                 ...   
60358  d01ad36f4033677a2e4883eee1a13af1   
60360  d01ad36f4033677a2e4883eee1a13af1   
60367  755183be01bfc739a9cac6673f1f2c1b   
60371  9f8dcf91c6164489adc6f0a51f7b07bf   
60372                               NaN   

      metadata.dump.dailymotion-domain-verification metadata.dump.og.domain  \
1                                               NaN                     NaN   
4                                               NaN                     NaN   
5                                               NaN                     NaN   
13                                              NaN                     NaN   
16                                              NaN                     NaN   
...                                             ...                     ...   
60358                                           NaN                     NaN   
60360                                           NaN                     NaN   
60367                                           NaN                     NaN   
60371                                           NaN                     NaN   
60372                                           NaN                     NaN   

      metadata.dump.com.silverpop.brandeddomains  \
1                                            NaN   
4                                            NaN   
5                                            NaN   
13                                           NaN   
16                                           NaN   
...                                          ...   
60358                                        NaN   
60360                                        NaN   
60367                                        NaN   
60371                                        NaN   
60372                                        NaN   

      metadata.dump.update-cart-domain  \
1                                  NaN   
4                                  NaN   
5                                  NaN   
13                                 NaN   
16                                 NaN   
...                                ...   
60358                              NaN   
60360                              NaN   
60367                              NaN   
60371                              NaN   
60372                              NaN   

      metadata.dump.preview-new-proration-domain  \
1                                            NaN   
4                                            NaN   
5                                            NaN   
13                                           NaN   
16                                           NaN   
...                                          ...   
60358                                        NaN   
60360                

In [104]:
#df_domain.shape

(60374, 19)

In [105]:
#df_domain_t=df_domain['metadata.dump.twitter.domain']

In [106]:
#df_domain_t =df_domain_t.dropna()

In [107]:
#df_domain_t.shape

(3499,)

### Only 3499 records in the original dataset have any domain names out of 60,000+ records. So, mapping up source-lean-flags.csv data into the original dataset will only give us much smaller data to play with


In [108]:
#df_domain_t.head(100)

1       www.13newsnow.com
4            www.aier.org
5                  rte.ie
18                 rte.ie
36                 rte.ie
              ...        
1501    www.thestreet.com
1511          www.bbc.com
1526               rte.ie
1542            Firstpost
1546         bitchute.com
Name: metadata.dump.twitter.domain, Length: 100, dtype: object

In [119]:
df_src_norm.head(3)

url  \
0  http://actionnewsjax.com/news/trending/amy-con...   
1  https://www.13newsnow.com/article/news/nation-...   
2  http://digg.com/digg-picks/link/kindles-video-...   

                                     extracted.title       extracted.date  \
0  Amy Coney Barrett hearing Day 2: Questioning b...                 None   
1  Vanessa Bryant wishes Kobe, Gigi could've seen...  2020-10-12T19:24:52   
2  Kindles, Video Doorbells And More Are On Sale ...                 None   

                                   extracted.authors  \
0  [Debbie Lord, Cox Media Group National Content...   
1                     [Author, Douglas Jones, Tegna]   
2                                     [Digg Editors]   

                                      extracted.text  \
0  “His judicial philosophy was straightforward: ...   
1  Vanessa Bryant wrote that she wished Kobe and ...   
2  👋 Welcome to Digg\n\nThanks for creating an ac...   

                                   metadata.keywords  \
0  [amy coney barrett, amy coney barrett hearing,...   
1                                                 []   
2                                                 []   

                                metadata.description metadata.language  \
0  Judge Amy Coney Barrett will face questions fr...                en   
1  After the Los Angeles Lakers won the NBA champ...                en   
2  <p>If you're a Prime member, Amazon has a nice...                en   

                                        metadata.url  \
0  https://www.actionnewsjax.com/news/trending/am...   
1  https://www.13newsnow.com/article/news/nation-...   
2  https://digg.com/digg-picks/link/kindles-video...   

  metadata.dump.distribution  ...  \
0                     global  ...   
1                        NaN  ...   
2                        NaN  ...   

  metadata.dump.The Echelon exercise bike is $100 off, down to $499, right now at Walmart — and it comes with a free 6-month membership  \
0                                                NaN                                                                                      
1                                                NaN                                                                                      
2                                                NaN                                                                                      

  metadata.dump.DC.source metadata.dump.DC.date.published  \
0                     NaN                             NaN   
1                     NaN                             NaN   
2                     NaN                             NaN   

  metadata.dump.webit_document_id metadata.dump.webit_document_name  \
0                             NaN                               NaN   
1                             NaN                               NaN   
2                             NaN                               NaN   

  metadata.dump.webit_cover_date metadata.dump.burntime  \
0                            NaN                    NaN   
1                            NaN                    NaN   
2                            NaN                    NaN   

  metadata.dump.Everything we know about Ariana Grande's sixth album  \
0                                                NaN                   
1                                                NaN                   
2                                                NaN                   

  metadata.dump.article_references metadata.dump.epdf_available  
0                              NaN                          NaN  
1                              NaN                          NaN  
2                              NaN                          NaN  

[3 rows x 2481 columns]

In [120]:
df_src_norm.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60374 entries, 0 to 60373
Columns: 2481 entries, url to metadata.dump.epdf_available
dtypes: bool(2), float64(251), object(2228)
memory usage: 1.1+ GB


In [111]:
#for col in df_src_norm.columns:
#    if 'metadata' in col:
#        del df_src_norm[col]

In [121]:
df_src_norm.shape

(60374, 2481)

In [122]:
df_src_norm_limited=df_src_norm.filter(items=['url', 'extracted.title', 'extracted.date', 'extracted.authors', 'extracted.text', 'source.url', 'source.sitename', 'flags.in_graph', 'flags.has_nlp', 'context.scraper', 'context.datetime', 'flags.skipped_graph', 'metadata.dump.twitter.domain'])

In [123]:
df_src_norm_limited.head(10)

url  \
0  http://actionnewsjax.com/news/trending/amy-con...   
1  https://www.13newsnow.com/article/news/nation-...   
2  http://digg.com/digg-picks/link/kindles-video-...   
3  http://sports.mynorthwest.com/1199798/clayton-...   
4  https://www.aier.org/article/lockdowns-have-ki...   
5  http://rte.ie/news/world/2020/1012/1171035-car...   
6  https://www.inforum.com/newsmd/coronavirus/671...   
7  https://www.postbulletin.com/news/crime-and-co...   
8  http://refinery29.com/de-de/2020/10/10080798/g...   
9  http://elnuevodia.com/noticias/mundo/videos/nu...   

                                     extracted.title  \
0  Amy Coney Barrett hearing Day 2: Questioning b...   
1  Vanessa Bryant wishes Kobe, Gigi could've seen...   
2  Kindles, Video Doorbells And More Are On Sale ...   
3  Seahawks Observations: John Clayton on 5-0 Haw...   
4  Lockdowns Have Killed What’s Left of the Unite...   
5  Teenager one step from becoming first millenni...   
6  Minnesota to require low-risk care homes be re...   
7  Rochester Police Department daily incident rep...   
8  „Gilmore Girls“: 16 berühmte Nebendarsteller*i...   
9  Nuevo misil balístico de Corea del Norte ¿Un d...   

                     extracted.date  \
0                              None   
1               2020-10-12T19:24:52   
2                              None   
3         2020-10-12T21:52:23+00:00   
4         2020-10-12T12:30:15+00:00   
5  2020-10-12T15:21:15.903000+01:00   
6                              None   
7                              None   
8               2020-10-10T00:00:00   
9                              None   

                                   extracted.authors  \
0  [Debbie Lord, Cox Media Group National Content...   
1                     [Author, Douglas Jones, Tegna]   
2                                     [Digg Editors]   
3              [John Clayton, Share, October, At Pm]   
4  [Peter C. Earle, October, Peter C. Earle Is An...   
5                                                 []   
6                  [Written, Paul John Scott, Oct -]   
7                      [Written, Andrew Link, Oct -]   
8  [Lindsay Denninger, Von Ineye Komonibo, Ineye ...   
9                                     [El Nuevo Día]   

                                      extracted.text  \
0  “His judicial philosophy was straightforward: ...   
1  Vanessa Bryant wrote that she wished Kobe and ...   
2  👋 Welcome to Digg\n\nThanks for creating an ac...   
3  The Seahawks’ 27-26 victory over the Minnesota...   
4                                                      
5  A British-born Italian teenager who dedicated ...   
6  ROCHESTER, Minn. — Stressing the need to balan...   
7  Wondering what the commotion was in your neigh...   
8  Ich spreche hier also nicht von Melissa McCart...   
9                                                      

                      source.url         source.sitename  flags.in_graph  \
0       http://actionnewsjax.com                    WJAX           False   
1      https://www.13newsnow.com           13newsnow.com           False   
2                http://digg.com                                   False   
3  http://sports.mynorthwest.com  sports.MyNorthwest.com           False   
4           https://www.aier.org                    AIER           False   
5                  http://rte.ie                  RTE.ie           False   
6        https://www.inforum.com                 INFORUM           False   
7   https://www.postbulletin.com           Post Bulletin           False   
8          http://refinery29.com                                   False   
9          http://elnuevodia.com            El Nuevo Día           False   

   flags.has_nlp context.scraper                  context.datetime  \
0          False  newspaperbasic  2020-10-12T22:03:11.021791+00:00   
1          False  newspaperbasic  2020-10-12T22:03:11.185397+00:00   
2          False  newspaperbasic  2020-10-12T22:03:11.585230+00:00   
3          False  newspaperbasic  

In [100]:
df_src_norm_limited['flags.in_graph'].value_counts()

False    35878
True     24496
Name: flags.in_graph, dtype: int64

In [124]:
df_src_norm_limited.shape

(60374, 13)

In [126]:
!pip install furl

You should consider upgrading via the '/opt/conda/bin/python3.7 -m pip install --upgrade pip' command.


In [150]:
# strip url of www and http

from furl import furl
def strip_url(url):
    #if "youtube.com/watch" in url:
    #    f = furl(url)
    #    try:
    #        v = f.args['v']
    #        url = f.remove(args=True, fragment=True).url
    #        url = furl(url).add(args={'v':v}).url
    #    except:
    #        pass
    #else:
    #    url = furl(url).remove(args=True, fragment=True).url
    
    if '://www.' in url:
        url = url.split('://www.',1)[1]
    elif '://' in url:
        url = url.split('://',1)[1]
    return url

In [151]:
strip_url('http://actionnewsjax.com')

'actionnewsjax.com'

In [157]:
# adding a new column domain_url by stripping source.url using strip_url function
df_src_norm_limited['domain_url']= df_src_norm_limited['source.url'].apply(strip_url)

In [156]:
df_src_norm_limited.head(10)

url  \
0  http://actionnewsjax.com/news/trending/amy-con...   
1  https://www.13newsnow.com/article/news/nation-...   
2  http://digg.com/digg-picks/link/kindles-video-...   
3  http://sports.mynorthwest.com/1199798/clayton-...   
4  https://www.aier.org/article/lockdowns-have-ki...   
5  http://rte.ie/news/world/2020/1012/1171035-car...   
6  https://www.inforum.com/newsmd/coronavirus/671...   
7  https://www.postbulletin.com/news/crime-and-co...   
8  http://refinery29.com/de-de/2020/10/10080798/g...   
9  http://elnuevodia.com/noticias/mundo/videos/nu...   

                                     extracted.title  \
0  Amy Coney Barrett hearing Day 2: Questioning b...   
1  Vanessa Bryant wishes Kobe, Gigi could've seen...   
2  Kindles, Video Doorbells And More Are On Sale ...   
3  Seahawks Observations: John Clayton on 5-0 Haw...   
4  Lockdowns Have Killed What’s Left of the Unite...   
5  Teenager one step from becoming first millenni...   
6  Minnesota to require low-risk care homes be re...   
7  Rochester Police Department daily incident rep...   
8  „Gilmore Girls“: 16 berühmte Nebendarsteller*i...   
9  Nuevo misil balístico de Corea del Norte ¿Un d...   

                     extracted.date  \
0                              None   
1               2020-10-12T19:24:52   
2                              None   
3         2020-10-12T21:52:23+00:00   
4         2020-10-12T12:30:15+00:00   
5  2020-10-12T15:21:15.903000+01:00   
6                              None   
7                              None   
8               2020-10-10T00:00:00   
9                              None   

                                   extracted.authors  \
0  [Debbie Lord, Cox Media Group National Content...   
1                     [Author, Douglas Jones, Tegna]   
2                                     [Digg Editors]   
3              [John Clayton, Share, October, At Pm]   
4  [Peter C. Earle, October, Peter C. Earle Is An...   
5                                                 []   
6                  [Written, Paul John Scott, Oct -]   
7                      [Written, Andrew Link, Oct -]   
8  [Lindsay Denninger, Von Ineye Komonibo, Ineye ...   
9                                     [El Nuevo Día]   

                                      extracted.text  \
0  “His judicial philosophy was straightforward: ...   
1  Vanessa Bryant wrote that she wished Kobe and ...   
2  👋 Welcome to Digg\n\nThanks for creating an ac...   
3  The Seahawks’ 27-26 victory over the Minnesota...   
4                                                      
5  A British-born Italian teenager who dedicated ...   
6  ROCHESTER, Minn. — Stressing the need to balan...   
7  Wondering what the commotion was in your neigh...   
8  Ich spreche hier also nicht von Melissa McCart...   
9                                                      

                      source.url         source.sitename  flags.in_graph  \
0       http://actionnewsjax.com                    WJAX           False   
1      https://www.13newsnow.com           13newsnow.com           False   
2                http://digg.com                                   False   
3  http://sports.mynorthwest.com  sports.MyNorthwest.com           False   
4           https://www.aier.org                    AIER           False   
5                  http://rte.ie                  RTE.ie           False   
6        https://www.inforum.com                 INFORUM           False   
7   https://www.postbulletin.com           Post Bulletin           False   
8          http://refinery29.com                                   False   
9          http://elnuevodia.com            El Nuevo Día           False   

   flags.has_nlp context.scraper                  context.datetime  \
0          False  newspaperbasic  2020-10-12T22:03:11.021791+00:00   
1          False  newspaperbasic  2020-10-12T22:03:11.185397+00:00   
2          False  newspaperbasic  2020-10-12T22:03:11.585230+00:00   
3          False  newspaperbasic  

In [160]:
#Now, join source-lean-flags dataframe with the main dataframe

dfinal = df_src_norm_limited.merge(df_src_flags, how='inner', left_on='domain_url', right_on='domain')

In [162]:
dfinal.head(10)

url  \
0  http://actionnewsjax.com/news/trending/amy-con...   
1  http://actionnewsjax.com/news/trending/john-mc...   
2  http://actionnewsjax.com/news/local/duval-coun...   
3  http://actionnewsjax.com/news/politics/biden-m...   
4  http://actionnewsjax.com/news/disney-reorg-fur...   
5  http://actionnewsjax.com/news/politics/barrett...   
6  http://actionnewsjax.com/news/local/duval-coun...   
7  http://actionnewsjax.com/news/local/clay-count...   
8  http://actionnewsjax.com/news/chinas-trade-gro...   
9  http://actionnewsjax.com/news/global-shares-mo...   

                                     extracted.title extracted.date  \
0  Amy Coney Barrett hearing Day 2: Questioning b...           None   
1  John McCain’s mother, Roberta McCain, dies at 108           None   
2  Police: Business owner shoots burglary suspect...           None   
3  Biden makes big push in Ohio, once seen as lon...           None   
4  Disney reorg to further bolster company's focu...           None   
5  Barrett vows fair approach as justice, Democra...           None   
6  As millions of Americans get ready for Amazon ...           None   
7  Clay County community gathers to honor lives o...           None   
8  China's trade growth accelerates in Sept; expo...           None   
9  Global shares mostly lower as virus concerns t...           None   

                                   extracted.authors  \
0  [Debbie Lord, Cox Media Group National Content...   
1  [Theresa Seiger, Cox Media Group National Cont...   
2  [Actionnewsjax.Com News Staff, Bridgette Matte...   
3  [Bill Barrow, Will Weissert, Updated, October,...   
4  [Tali Arbel, Updated, October, - Pm, Kelli Dug...   
5  [Mark Sherman, Lisa Mascaro, Mary Clare Jaloni...   
6  [Dani Bozzini, Action News Jax, Updated, Octob...   
7  [Ryan Nelson, Action News Jax, Updated, Octobe...   
8  [Joe Mcdonald, Updated, October, - Am, Michell...   
9  [Elaine Kurtenbach, Updated, October, - Am, Mi...   

                                      extracted.text  \
0  “His judicial philosophy was straightforward: ...   
1  Known for her feisty and energetic spirit, Rob...   
2  “He was hiding under all these and bleeding ri...   
3  The Biden campaign has increased advertising i...   
4  The coronavirus pandemic has hit box-office re...   
5  Testifying from her office because of the pand...   
6  When the pandemic began, Native Sun shifted it...   
7  Rowley died at the scene, and Kennedy died at ...   
8  China became the first major economy to reboun...   
9  Optimism over hopes for new aid for the U.S. e...   

                 source.url source.sitename  flags.in_graph  flags.has_nlp  \
0  http://actionnewsjax.com            WJAX           False          False   
1  http://actionnewsjax.com            WJAX           False          False   
2  http://actionnewsjax.com            WJAX           False          False   
3  http://actionnewsjax.com            WJAX           False          False   
4  http://actionnewsjax.com            WJAX            True          False   
5  http://actionnewsjax.com            WJAX           False          False   
6  http://actionnewsjax.com            WJAX           False          False   
7  http://actionnewsjax.com            WJAX           False          False   
8  http://actionnewsjax.com            WJAX           False          False   
9  http://actionnewsjax.com            WJAX            True          False   

  context.scraper  ... flags.skipped_graph metadata.dump.twitter.domain  \
0  newspaperbasic  ...                 NaN                          NaN   
1  newspaperbasic  ...                 NaN                          NaN   
2  newspaperbasic  ...                 NaN                          NaN   
3  newspaperbasic  ...                 NaN                          NaN   
4  newspaperbasic  ...                 NaN                          NaN   
5  newspaperbasic  ...                True                          NaN   
6  newspaperbasic  ...          

In [163]:
dfinal.shape

(51102, 21)

In [51]:
#df_src_norm_limited=df_src_norm.filter(items=['url','source.sitename','extracted.title', 'extracted.date', 'extracted.text', 'metadata.keywords', 'metadata.description'])

In [58]:
#df_src_norm_limited.shape

In [59]:
#df_src_norm_limited.head(10)

In [60]:
#df_src_norm_limited['metadata.description'][0]

In [61]:
#df_src_norm_limited['extracted.text'][0]

In [ ]:
#df_src_norm_limited['url'][0]

In [ ]:
#df_src_norm_limited.describe

## Create Datasets for further analysis


In [62]:
#extracted_text=df_src_norm_limited['extracted.text']
#extracted_text.dropna()


In [63]:
#extracted_text.shape

# Explore some transformers

### https://github.com/UKPLab/sentence-transformers

In [17]:
!pip install -U sentence-transformers

     |████████████████████████████████| 59 kB 3.6 MB/s eta 0:00:011
     |████████████████████████████████| 1.1 MB 7.4 MB/s eta 0:00:01
     |████████████████████████████████| 3.0 MB 35.6 MB/s eta 0:00:01
  Created wheel for sentence-transformers: filename=sentence_transformers-0.3.7.2-py3-none-any.whl size=91186 sha256=5311dc98f80cdddd2f83dd7f4a5f22db20c5f542aa9ddc5705bfc13d36e6977b
  Stored in directory: /root/.cache/pip/wheels/df/42/15/b8329fd622ed2e9ee15b00314a387aa7d70e402d144349d540
Successfully built sentence-transformers
  Attempting uninstall: tokenizers
    Found existing installation: tokenizers 0.7.0
    Uninstalling tokenizers-0.7.0:
      Successfully uninstalled tokenizers-0.7.0
  Attempting uninstall: transformers
    Found existing installation: transformers 2.10.0
    Uninstalling transformers-2.10.0:
      Successfully uninstalled transformers-2.10.0
You should consider upgrading via the '/opt/conda/bin/python3.7 -m pip install --upgrade pip' command.


In [66]:
from sentence_transformers import SentenceTransformer
model = SentenceTransformer('distilbert-base-nli-stsb-mean-tokens')

#Our sentences we like to encode
sentences = ['This framework generates embeddings for each input sentence',
    'Sentences are passed as a list of string.', 
    'The quick brown fox jumps over the lazy dog.']

#Sentences are encoded by calling model.encode()
sentence_embeddings = model.encode(sentences)

#Print the embeddings
for sentence, embedding in zip(sentences, sentence_embeddings):
    print("Sentence:", sentence)
    #print("Embedding:", embedding)
    print("")

wandb: WARNING W&B installed but not logged in.  Run `wandb login` or set the WANDB_API_KEY env variable.


Sentence: This framework generates embeddings for each input sentence

Sentence: Sentences are passed as a list of string.

Sentence: The quick brown fox jumps over the lazy dog.



In [64]:
#extracted_text.head(10)

In [24]:
from sentence_transformers import SentenceTransformer, util
model = SentenceTransformer('distilbert-base-nli-stsb-mean-tokens')

sentences = ['A man is eating food.',
          'A man is eating a piece of bread.',
          'The girl is carrying a baby.',
          'A man is riding a horse.',
          'A woman is playing violin.',
          'Two men pushed carts through the woods.',
          'A man is riding a white horse on an enclosed ground.',
          'A monkey is playing drums.',
          'Someone in a gorilla costume is playing a set of drums.'
          ]

#Encode all sentences
embeddings = model.encode(sentences)

#Compute cosine similarity between all pairs
cos_sim = util.pytorch_cos_sim(embeddings, embeddings)

#Add all pairs to a list with their cosine similarity score
all_sentence_combinations = []
for i in range(len(cos_sim)-1):
    for j in range(i+1, len(cos_sim)):
        all_sentence_combinations.append([cos_sim[i][j], i, j])

#Sort list by the highest cosine similarity score
all_sentence_combinations = sorted(all_sentence_combinations, key=lambda x: x[0], reverse=True)

print("Top-5 most similar pairs:")
for score, i, j in all_sentence_combinations[0:5]:
    print("{} \t {} \t {:.4f}".format(sentences[i], sentences[j], cos_sim[i][j]))

Top-5 most similar pairs:
A man is eating food. 	 A man is eating a piece of bread. 	 0.7210
A monkey is playing drums. 	 Someone in a gorilla costume is playing a set of drums. 	 0.6435
A man is riding a horse. 	 A man is riding a white horse on an enclosed ground. 	 0.6343
A man is eating food. 	 A man is riding a horse. 	 0.2840
A man is eating a piece of bread. 	 A monkey is playing drums. 	 0.2451


In [65]:
#metadata_description=df_src_norm_limited['metadata.description']